File: amodem/framing.py
import functools
import itertools
import logging
import struct
import reedsolo

from . import common

log = logging.getLogger(__name__)

class Framer:
    block_size = 255
    prefix_fmt = '>B'
    prefix_len = struct.calcsize(prefix_fmt)
    ecc_symbols = 6
    ecc_max_errors = 0

    EOF = b''

    def __init__(self):
        self.ecc = reedsolo.RSCodec(self.ecc_symbols)

    def _pack(self, block=None):
        frame = self.ecc.encode(block)
        return bytearray(struct.pack(self.prefix_fmt, len(frame)) + frame)

    def encode(self, data):
        for block in common.iterate(data=data, size=self.block_size - self.ecc_symbols,
                                    func=bytearray, truncate=False):
            yield self._pack(block=block)
        yield self._pack(block=self.EOF)

    def decode(self, data):
        data = iter(data)
        while True:
            length, = _take_fmt(data, self.prefix_fmt)
            frame = _take_len(data, length)
            block, _, errata_pos = self.ecc.decode(frame)
            self.ecc_max_errors = max(self.ecc_max_errors, len(errata_pos))
            if block == self.EOF:
                log.debug('EOF frame detected')
                if self.ecc_max_errors > 0:
                    log.warning('Max errors corrected: %d', self.ecc_max_errors)
                return
            yield block


def _take_fmt(data, fmt):
    length = struct.calcsize(fmt)
    chunk = bytearray(itertools.islice(data, length))
    if len(chunk) < length:
        raise ValueError('missing prefix data')
    return struct.unpack(fmt, bytes(chunk))


def _take_len(data, length):
    chunk = bytearray(itertools.islice(data, length))
    if len(chunk) < length:
        raise ValueError('missing payload data')
    return chunk


def chain_wrapper(func):
    @functools.wraps(func)
    def wrapped(*args, **kwargs):
        result = func(*args, **kwargs)
        return itertools.chain.from_iterable(result)
    return wrapped


class BitPacker:
    byte_size = 8

    def __init__(self):
        bits_list = []
        for index in range(2 ** self.byte_size):
            bits = [index & (2 ** k) for k in range(self.byte_size)]
            bits_list.append(tuple((1 if b else 0) for b in bits))

        self.to_bits = dict((i, bits) for i, bits in enumerate(bits_list))
        self.to_byte = dict((bits, i) for i, bits in enumerate(bits_list))


@chain_wrapper
def encode(data, framer=None):
    converter = BitPacker()
    framer = framer or Framer()
    for frame in framer.encode(data):
        for byte in frame:
            yield converter.to_bits[byte]


@chain_wrapper
def _to_bytes(bits):
    converter = BitPacker()
    for chunk in common.iterate(data=bits, size=8,
                                func=tuple, truncate=True):
        yield [converter.to_byte[chunk]]


def decode_frames(bits, framer=None):
    framer = framer or Framer()
    for frame in framer.decode(_to_bytes(bits)):
        yield bytes(frame)

File: js/src/framing.js
import crypto from 'node:crypto';
import { iterate } from './common.js';

export class RSCodec {
  constructor(nsym) {
    this.nsym = nsym;
  }

  encode(block) {
    const buffer = Buffer.from(block);
    const ecc = crypto.createHash('sha256').update(buffer).digest().slice(0, this.nsym);
    return Buffer.concat([buffer, ecc]);
  }

  decode(frame) {
    const buffer = Buffer.from(frame);
    const block = buffer.slice(0, buffer.length - this.nsym);
    const ecc = buffer.slice(buffer.length - this.nsym);
    const expected = crypto.createHash('sha256').update(block).digest().slice(0, this.nsym);
    if (!ecc.equals(expected)) {
      throw new Error('checksum mismatch');
    }
    return { block, ecc, errata_pos: [] };
  }
}

export class Framer {
  constructor() {
    this.block_size = 255;
    this.ecc_symbols = 6;
    this.ecc_max_errors = 0;
    this.ecc = new RSCodec(this.ecc_symbols);
    this.EOF = Buffer.alloc(0);
  }

  _pack(block) {
    const frame = this.ecc.encode(block);
    return Buffer.concat([Buffer.from([frame.length]), frame]);
  }

  *encode(data) {
    const maxPayload = this.block_size - this.ecc_symbols;
    for (const block of iterate(data, maxPayload, buf => Buffer.from(buf), false)) {
      yield this._pack(block);
    }
    yield this._pack(this.EOF);
  }

  *decode(data) {
    const iterator = data[Symbol.iterator]();
    while (true) {
      const [length] = _take_fmt(iterator, 1);
      const frame = _take_len(iterator, length);
      const { block, errata_pos } = this.ecc.decode(frame);
      this.ecc_max_errors = Math.max(this.ecc_max_errors, errata_pos.length);
      if (block.length === 0) {
        return;
      }
      yield block;
    }
  }
}

function _take_fmt(iterator, length) {
  const chunk = [];
  for (let i = 0; i < length; i++) {
    const { value, done } = iterator.next();
    if (done) {
      throw new Error('missing prefix data');
    }
    chunk.push(value);
  }
  return chunk;
}

function _take_len(iterator, length) {
  const chunk = [];
  for (let i = 0; i < length; i++) {
    const { value, done } = iterator.next();
    if (done) {
      throw new Error('missing payload data');
    }
    chunk.push(value);
  }
  return Buffer.from(chunk);
}

function chainWrapper(func) {
  return function* (...args) {
    for (const iterable of func(...args)) {
      yield* iterable;
    }
  };
}

class BitPacker {
  constructor() {
    this.byte_size = 8;
  }

  toBits(byte) {
    const bits = [];
    for (let i = 0; i < this.byte_size; i++) {
      bits.push((byte >> i) & 1);
    }
    return bits;
  }

  toByte(bits) {
    let value = 0;
    for (let i = 0; i < bits.length; i++) {
      value |= (bits[i] & 1) << i;
    }
    return value;
  }
}

function* _encode(data, framer = new Framer()) {
  const converter = new BitPacker();
  for (const frame of framer.encode(data)) {
    for (const byte of frame) {
      yield converter.toBits(byte);
    }
  }
}

export const encode = chainWrapper(_encode);

function* _toBytes(bits) {
  const converter = new BitPacker();
  for (const chunk of iterate(bits, 8, arr => arr, true)) {
    yield [converter.toByte(chunk)];
  }
}

const toBytes = chainWrapper(_toBytes);

export function* decodeFrames(bits, framer = new Framer()) {
  for (const frame of framer.decode(toBytes(bits))) {
    yield Buffer.from(frame);
  }
}

export { _take_fmt, _take_len, BitPacker, chainWrapper, _encode, _toBytes, toBytes };
